applications/TrilinosApplication/external_includes/amesos2_solver.h

```diff
@@ -0,0 +1,251 @@
+            // do nothing here.
+            // Leave full control to the user through the "trilinos_amesos2_parameter_list"
+            // and the "amesos2_solver_type"
+        }
+
+        else {
+            KRATOS_ERROR << "The solver type specified: \"" << solver_type << "\" is not supported";
+        }
+
+        settings.ValidateAndAssignDefaults(default_settings);
+
+        //assign the amesos parameter list, which may contain parameters IN TRILINOS INTERNAL FORMAT to mParameterList
+        //NOTE: this will OVERWRITE PREVIOUS SETTINGS TO GIVE FULL CONTROL
+        TrilinosSolverUtilities::SetTeuchosParameters(settings["trilinos_amesos2_parameter_list"], mParameterList);
+
+        mSolverName = settings["amesos2_solver_type"].GetString();
+
+        KRATOS_ERROR_IF_NOT(HasSolver(mSolverName)) << "attempting to use Amesos solver \"" << mSolverName
+            << "\" unfortunately the current compilation of Trilinos does not include it" << std::endl;
+    }
+
+    /// Constructor with solver-name and Teuchos::ParameterList.
+    Amesos2Solver(const std::string& SolverName, Teuchos::ParameterList& rParameterList)
+    {
+        mParameterList = rParameterList;
+        mSolverName = SolverName;
+
+        KRATOS_ERROR_IF_NOT(HasSolver(mSolverName)) << "attempting to use Amesos solver \"" << mSolverName
+            << "\" unfortunately the current compilation of Trilinos does not include it" << std::endl;
+    }
+
+    /// Copy constructor.
+    Amesos2Solver(const Amesos2Solver& Other) = delete;
+
+    /// Destructor.
+    ~Amesos2Solver() override = default;
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    /// Assignment operator.
+    Amesos2Solver& operator=(const Amesos2Solver& Other) = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * Normal solve method.
+     * Solves the linear system Ax=b and puts the result on SystemVector& rX.
+     * rX is also th initial guess for iterative methods.
+     * @param rA. System matrix
+     * @param rX. Solution vector.
+     * @param rB. Right hand side vector.
+     */
+    bool Solve(SparseMatrixType& rA, VectorType& rX, VectorType& rB) override
+    {
+        // implemented following this example:
+        // https://github.com/trilinos/Trilinos/blob/master/packages/amesos2/example/SimpleSolve_File.cpp
+
+        KRATOS_TRY
+
+        // it is not strictly necessary to use rcp, however skipping it could not yet be achieved
+        // in any case doesn't hurt to have it
+        using Teuchos::RCP;
+        using Teuchos::rcp;
+
+        // Amesos2 is missing specializations for FE-matrices and vectors
+        // hence here we need to use the corresponding baseclasses
+        // see "Amesos2_MatrixTraits.hpp" and "Amesos2_VectorTraits.hpp"
+        using MAT = Epetra_CrsMatrix;   // baseclass of "SparseMatrixType" aka "Epetra_FECrsMatrix"
+        using MV  = Epetra_MultiVector; // baseclass of "VectorType"       aka "Epetra_FEVector
```

Comment:
For the record, I tried to provide manually the MatrixAdapter for the FE-Matrix & Vector, but was not successful (see https://docs.trilinos.org/dev/packages/amesos2/doc/html/classAmesos2_1_1MatrixAdapter.html)

in the future when we switch to Kratos Matrices I think we can take another look, we can probably make it work without copies by providing a suitable adapter